### PR TITLE
Add assign-class-label mutating webhook

### PR DESCRIPTION
**Pull request by IsaiahStapleton**
Closes: https://github.com/nerc-project/operations/issues/637

If you want to test this yourself. It is currently deployed in both prod and test cluster. Just add yourself to a group ee440, cs210, or cs506. Request to create a pod in rhods-notebooks namespace (from jupyter tile) and check your pod labels afterward and see if you have a class label. You can also check the logs of the webhook here: https://console.apps.shift.nerc.mghpcc.org/k8s/ns/rhods-notebooks/pods/assign-class-label-webhook-7f6db498fd-rwdmt/logs which should show it assigning label to your pod.

README pending

### CODE REVIEW

File: docker/src/python/assign-class-label/mutate.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                "uid": uid,
+                "allowed": True,
+                "status": {"message": "No class label assigned."}
+            }
+        })
+    
+    # Generate JSON Patch to add class label
+    patch = [{
+        "op": "add",
+        "path": "/metadata/labels/class",
```
Comment (larsks) [inline docker/src/python/assign-class-label/mutate.py:85]: I would consider using a "qualifed" label instead of simple label like `class` (because for short common words there's a good chance that someone might try using the same label to mean something else).  That's why, for example we label namespaces in the production cluster with `nerc.mghpcc.org/project` instead of simply `project`.

File: docker/src/python/assign-class-label/mutate.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+            "uid": uid,
+            "allowed": True,
+            "patchType": "JSONPatch",
+            "patch": patch_base64
+        }
+    })
+
+if __name__ == "__main__":
+    logging.basicConfig(level="INFO")
+    webhook.run(host='0.0.0.0', port=8443)
```
Comment (larsks) [inline docker/src/python/assign-class-label/mutate.py:106]: If you remove this chunk of code you can `import` this module from a test or from the repl, which is often helpful. You can always run it locally with `flask --app wsgi:webhook run` if necessary.

File: docker/src/python/assign-class-label/mutate.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+            # Check if group has no users
+            if not group_users:
+                LOG.warning(f"Group {group} has no users or users attribute is not a list.")
+                continue
+
+            # Compare users in the groups (classes) with the pod user
+            if pod_user in group_users:
+                LOG.info(f"Assigning class label: {group} to user {pod_user}")
+                return group
```
Comment (larsks) [inline docker/src/python/assign-class-label/mutate.py:47]: You can access `group_obj.users` directly without converting it to a `dict` first. `group_obj.users` will return `None` if the attribute does not exist (that is, it works pretty much exactly like the `get` method on a `dict`).

```suggestion
            # Check if group has no users
            if not group_obj.users:
                LOG.warning(f"Group {group} has no users or users attribute is not a list.")
                continue

            # Compare users in the groups (classes) with the pod user
            if pod_user in group_obj.users:
                LOG.info(f"Assigning class label: {group} to user {pod_user}")
                return group
```

File: webhooks/assign-class-label/deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+kind: Deployment
+apiVersion: apps/v1
+metadata:
+  name: assign-class-label-webhook
+  labels:
+    app: assign-class-label
```
Comment (larsks) [inline webhooks/assign-class-label/deployment.yaml:6]: If you remove your labels in `metadata.labels` and `spec.template.metadata.labels`, and you remove `spec.selector`, and set them in `kustomization.yaml` instead (...which you are already doing), kustomize will take care of setting everything correctly.

File: webhooks/assign-class-label/service.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: assign-class-label-webhook
+spec:
+  selector:                  
+    app: assign-class-label-webhook
```
Comment (larsks) [inline webhooks/assign-class-label/service.yaml:7]: You can remove the `selector` and let kustomize manage it for you.

File: docker/src/python/assign-class-label/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,11 @@
...
+
+WORKDIR /app
+
+COPY . /app
+
+RUN pip install -r requirements.txt
+
+EXPOSE 8443
+
+CMD ["gunicorn", "wsgi:webhook", "--log-level=info", "--workers", "3", "--bind", "0.0.0.0:8443", "--keyfile", "/certs/webhook.key", "--certfile", "/certs/webhook.crt"]
```
Comment (larsks) [inline docker/src/python/assign-class-label/Dockerfile:11]: You should not be running an http service on port `8443` (because that's confusing and people will expect it to be https).

File: webhooks/assign-class-label/webhook-config.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
...QzRIV1c0QW4vS1VoVW5pNkpWbHhjaE02dWJMeU5sNjBmRVRERWJBeHgzWVUKa1RHa01JNGxWS0VlVGlpcFJVQUdwQjg0VXBEbi82cnpwQUJqT2xoakFmMUNZcnhzSVJvc1o2L1ZVcG9veTdOZApiYllIdlFTMmd1N2IwODAyS2YrWko2bUpPbGlIcUdtVjRqVlRJK0RPem15RXlhMmhwSXpHMnZOQ0JIcDJVZXN2CkdaMXJ0dE8xOWs3K1AyQXdoZFVUMk14Z2NEeUdaSjF5N2ZuR25JbXczRGtEL2FKWUp3bklHaWhaMnNSZTlsbjgKbGVsRVN3YWx3S2NPTlZRcnB3QlJzdC8ySXdhdlpyd1dNQVg0cUpEeks3dWJyc3FJdkZZcUh6MENBd0VBQWFPQgpqekNCakRBZEJnTlZIUTRFRmdRVXpDWUswU2tkbm02S0lNbnFiSGF1QnlSOWxrb3dId1lEVlIwakJCZ3dGb0FVCnpDWUswU2tkbm02S0lNbnFiSGF1QnlSOWxrb3dEd1lEVlIwVEFRSC9CQVV3QXdFQi96QTVCZ05WSFJFRU1qQXcKZ2k1aGMzTnBaMjR0WTJ4aGMzTXRiR0ZpWld3dGQyVmlhRzl2YXk1eWFHOWtjeTF1YjNSbFltOXZhM011YzNaagpNQTBHQ1NxR1NJYjNEUUVCQ3dVQUE0SUJBUUI3eVpGM1k5ZWFkTDR5Y1R3cHJsem5JT2IreUJncGxkOFdlNndqCisvWE9XVnlXb09rM1JYT0VadWoxUE8xcWpHY0lJOFpQc05xamJmNVhOd1hkaUlsSm1KdDJaQVFWWkpqY0dkMmQKS2RRNXF1Wk44Z2ZFS0pwWHJUVWlIZ3VQQjdCNmd5dGVBYWlyaXl4RDllZmgzcG1HTGJaU0xscXlRcVBrUFZ6UwpUeXBIbDNPVHllaE8wKy82MWh0MnE1TVpYamd4Rk1CQTlqNXZKd2JUMlVxWFllSGFOOTlqcU0rTW10Zks2NUUxCjkxQUY2bndmQys0Z1FNS0tUQ0hpVkFWb09KakJBVjZyOXkydSs5a0xyNE9LUkFaWEF4alI3dFl1YitsQmJYZUsKSHpOdnBYajBLZmpBaG9BNkdPUW1MVG01UGRNZjN2dWhlS1FFdmp1L3lldzZTMlk5Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
```
Comment (larsks) [inline webhooks/assign-class-label/webhook-config.yaml:12]: You should not need to provide this here; from the docs:

```
$ kubectl explain MutatingWebhookConfiguration.webhooks.clientConfig.caBundle
GROUP:      admissionregistration.k8s.io
KIND:       MutatingWebhookConfiguration
VERSION:    v1

FIELD: caBundle <string>

DESCRIPTION:
    `caBundle` is a PEM encoded CA bundle which will be used to validate the
    webhook's server certificate. If unspecified, system trust roots on the
    apiserver are used.
```

The "system trust roots on the apiserver" should be sufficient.

I bring this up because we would need a different `caBundle` on every server on which this ever gets deployed, whereas if we can remove it from the manifest, we can use the same manifest everywhere.

**Review by larsks — CHANGES_REQUESTED** [review]
I've left some comments on the code; some I feel more strongly about than others. Feel free to push back if something isn't clear.

I would like to see some tests (and a workflow to run them automatically on new pull requests).

**Comment by larsks** [comment]
Also, the organization of the `docker/` directory seems a little suspect (you've got a top-level `docker/Dockerfile`, but then you have `docker/src/python/assign-class-label/Dockerfile underneath it). I would at least separate your two Python apps, maybe:

```
docker/
├── assign-class-label
│   ├── Dockerfile
│   ├── mutate.py
│   ├── requirements.txt
│   └── wsgi.py
└── group-sync
    ├── Dockerfile
    ├── group-sync.py
    └── requirements.txt
```

File: docker/src/python/assign-class-label/mutate.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+webhook = Flask(__name__)
+
+LOG = logging.getLogger(__name__)
+logging.basicConfig(level=logging.INFO)
+
+try:
+    config.load_incluster_config()
+except config.ConfigException as e:
+    LOG.error("Could not configure Kubernetes client: %s", str(e))
+    exit(1)
```
Comment (larsks) [inline docker/src/python/assign-class-label/mutate.py:18]: To make this easier to run locally, rewrite this to use `config.load_kube_config()` if `config.load_incluster_config()` fails. Something like:


```suggestion
try:
    config.load_incluster_config()
except config.ConfigException:
    try:
        config.load_kube_config()
    except config.ConfigException:
        LOG.error("Could not configure Kubernetes client")
        exit(1)
```

File: docker/src/python/assign-class-label/mutate.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    return None
+
+@webhook.route('/mutate', methods=['POST'])
+def mutate_pod():
+    # Grab pod for mutation 
+    request_info = request.get_json()
+    uid = request_info["request"]["uid"]
+    pod = request_info["request"]["object"]
+
+    groups = os.environ.get("GROUPS").split(",")
```
Comment (larsks) [inline docker/src/python/assign-class-label/mutate.py:61]: You should call this something other than `GROUPS` (because `GROUPS` is the name of a standard shell variable, which may cause confusion when running things locally).

File: docker/src/python/assign-class-label/mutate.py [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+
+k8s_client = client.ApiClient()
+dyn_client = DynamicClient(k8s_client)
+
+
+def assign_class_label(pod, groups):
+    # Extract pod metadata
+    pod_metadata = pod.get('metadata', {})
+    pod_labels = pod_metadata.get('labels', {})
+    pod_user = pod_labels.get("opendatahub.io/user", None)
```
Comment (larsks) [inline docker/src/python/assign-class-label/mutate.py:28]: If `pod_user` is `None` we should probably just stop here.

**Comment by IsaiahStapleton** [comment]
@larsks Please review the changes to this PR when you get a chance.

File: .github/workflows/assign-class-label.yaml [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+          ruff format
+
+      - name: Run ruff check (linting)
+        run: |
+          ruff check --fix
+
+      - name: Run tests
+        working-directory: ./container-images/assign-class-label/
+        run: |
+          PYTHONPATH=. pytest tests/
```
Comment (larsks) [inline .github/workflows/assign-class-label.yaml:41]: You should not need to manipulate `PYTHONPATH` in order to run your tests. Instead, create an empty file `container-images/assign-class-label/conftest.py`; this will allow pytest to locate your code.

```suggestion
          pytest tests/
```

File: .github/workflows/assign-class-label.yaml [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+          pip install -r requirements.txt
+          pip install -r tests/requirements.txt
+
+      - name: Run ruff format (formatting)
+        run: |
+          ruff format
+
+      - name: Run ruff check (linting)
+        run: |
+          ruff check --fix
```
Comment (larsks) [inline .github/workflows/assign-class-label.yaml:36]: Note that you are running `ruff` from the top-level directory, rather than from the `container-images/assign-class-label` directory. That means you will potentially be checking files other than in the `assign-class-label` directory.

If that's intentional, you should move this into a different workflow. Otherwise, you should probably add a `working-directory` option to these tasks as well. Alternately, set a default for the whole workflow in [`defaults.run.working-directory`](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#defaultsrunworking-directory).

File: .github/workflows/assign-class-label.yaml [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: '3.12'  
+
+      - name: Install dependencies
+        working-directory: ./container-images/assign-class-label/
+        run: |
+          pip install -r requirements.txt
+          pip install -r tests/requirements.txt
```
Comment (larsks) [inline .github/workflows/assign-class-label.yaml:28]: Typically, this file would live in the same directory as `requirements.txt` and would be called `test-requirements.txt`.

File: container-images/assign-class-label/Dockerfile [hunk 14]
```diff
@@ -0,0 +1,13 @@
+FROM python:3.9-slim
+
+WORKDIR /app/
+
+COPY requirements.txt /app/
```
Comment (larsks) [inline container-images/assign-class-label/Dockerfile:5]: Since you've just performed `WORKDIR /app` in the previous step, you can just copy into the current directory instead:

```suggestion
COPY requirements.txt ./
```

File: container-images/assign-class-label/Dockerfile [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+
+RUN pip install -r requirements.txt
+
+COPY . /app/
```
Comment (larsks) [inline container-images/assign-class-label/Dockerfile:9]: And similarly:

```suggestion
COPY . ./
```

File: container-images/assign-class-label/Dockerfile [hunk 16]
```diff
@@ -0,0 +1,13 @@
+FROM python:3.9-slim
```
Comment (larsks) [inline container-images/assign-class-label/Dockerfile:1]: Is there a specific reason you're using an older Python? If not, I think you should use the current release.

```suggestion
FROM python:3.12-slim
```

File: container-images/assign-class-label/mutate.py [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+    object: PodObject
+
+
+class AdmissionReview(BaseModel):
+    request: AdmissionRequest
+
+
+def decode_pod_user(pod_user: str) -> str:
+    user = pod_user.replace('-40', '@').replace('-2e', '.')
+    return user
```
Comment (larsks) [inline container-images/assign-class-label/mutate.py:34]: Since you're not using the return value inside this function, you can drop the local variable:

```suggestion
    return pod_user.replace('-40', '@').replace('-2e', '.')
```

File: container-images/assign-class-label/mutate.py [hunk 18]
```diff
@@ -0,0 +1,180 @@
...
+
+# Get users of a given group
+def get_group_members(group_resource: Any, group_name: str) -> List[str]:
+    group_obj = group_resource.get(name=group_name)
+    return group_obj.users
+
+
+def assign_class_label(
+    pod: Dict[str, Any], groups: List[str], dyn_client: DynamicClient
+) -> Optional[str]:
```
Comment (larsks) [inline container-images/assign-class-label/mutate.py:60]: Or (with Python 3.12):

```suggestion
def assign_class_label(
    pod: dict[str, Any], groups: list[str], dyn_client: DynamicClient
) -> str | None:
```

File: container-images/assign-class-label/mutate.py [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+    try:
+        pod_metadata = pod.get('metadata', {})
+        pod_labels = pod_metadata.get('labels', {})
+        pod_user = pod_labels.get('opendatahub.io/user', None)
+    except AttributeError as e:
+        LOG.error(f'Error extracting pod information: {e}')
+        return None
+    
+    if pod_user is not None:
+        pod_user = decode_pod_user(pod_user)
```
Comment (larsks) [inline container-images/assign-class-label/mutate.py:71]: If pod user **is** `None`, shouldn't we just return?

```suggestion
    if pod_user is None:
        return None
        
    pod_user = decode_pod_user(pod_user)
```

File: container-images/assign-class-label/mutate.py [hunk 20]
```diff
@@ -0,0 +1,180 @@
...
+                        'response': {
+                            'uid': request.json.get('request', {}).get('uid', ''),
+                            'allowed': False,
+                            'status': {'message': f'Invalid request: {e}'},
+                        },
+                    }
+                ),
+                400,
+                {'content-type': 'application/json'},
+            )
```
Comment (larsks) [inline container-images/assign-class-label/mutate.py:126]: You should be using a pydantic model to generate this response, rather than manually constructing the response structure. This provides the same validation for your output that you get when reading input in the previous step.

File: container-images/assign-class-label/mutate.py [hunk 21]
```diff
@@ -0,0 +1,180 @@
...
+            return jsonify(
+                {
+                    'apiVersion': 'admission.k8s.io/v1',
+                    'kind': 'AdmissionReview',
+                    'response': {
+                        'uid': uid,
+                        'allowed': True,
+                        'status': {'message': 'No class label assigned.'},
+                    },
+                }
```
Comment (larsks) [inline container-images/assign-class-label/mutate.py:149]: See above :)

File: container-images/assign-class-label/mutate.py [hunk 22]
```diff
@@ -0,0 +1,180 @@
...
+            {
+                'apiVersion': 'admission.k8s.io/v1',
+                'kind': 'AdmissionReview',
+                'response': {
+                    'uid': uid,
+                    'allowed': True,
+                    'patchType': 'JSONPatch',
+                    'patch': patch_base64,
+                },
+            }
```
Comment (larsks) [inline container-images/assign-class-label/mutate.py:177]: And also here.

File: container-images/group-sync/group-sync.py
Comment (larsks) [inline container-images/group-sync/group-sync.py:1]: I would move these changes to a separate pull request, since they're not related to your work on the `assign-class-labels` webhook.

**Review by larsks — COMMENTED** [review]
This is a first pass review; I'll do a second pass tomorrow where I take a closer look at the code itself.

**Review by larsks — APPROVED** [review]

**Review by DanNiESh — APPROVED** [review]
